src/Databases/DatabaseAtomic.cpp

```diff
@@ -160,6 +160,43 @@ void DatabaseAtomic::dropTableImpl(ContextPtr local_context, const String & tabl
     DatabaseCatalog::instance().enqueueDroppedTableCleanup(table->getStorageID(), table, table_metadata_path_drop, sync);
 }
 
+void DatabaseAtomic::dropDetachedTable(ContextPtr local_context, const String & table_name, const bool sync)
+{
+    waitDatabaseStarted();
+
+    assert(fs::exists(getObjectMetadataPath(table_name)));
+
+    const String table_metadata_path = getObjectMetadataPath(table_name);
+    const UUID uuid_table = GetTableUUIDFromDetachedMetadata(local_context, table_metadata_path);
+
+    waitDetachedTableNotInUse(uuid_table);
+
+    const StorageID storage_id{getDatabaseName(), table_name, uuid_table};
+    const String table_metadata_path_drop = DatabaseCatalog::instance().getPathForDroppedMetadata(storage_id);
+
+    {
+        std::lock_guard lock(mutex);
+
+        auto txn = local_context->getZooKeeperMetadataTransaction();
+        if (txn && !local_context->isInternalSubquery())
+            txn->commit();
+
+        fs::rename(table_metadata_path, table_metadata_path_drop);
+        LOG_TRACE(log, "Rename {} to {} for removing.", table_metadata_path, table_metadata_path_drop);
+
+        DatabaseCatalog::instance().removeUUIDMappingFinally(uuid_table);
+        table_name_to_path.erase(table_name);
+    }
+
+    tryRemoveSymlink(table_name);
```

Comment:
What if the table does not have data on the disk?
see `dropTableImpl`:
https://github.com/ClickHouse/ClickHouse/blob/e408c187c7462fc5fc4b661f71f2194a4b9e3cec/src/Databases/DatabaseAtomic.cpp#L155-L156

Reply:
We will try to remove dir that doesn't exsit in exception-safe method. I think its ok. At least I don't know problem right now.